File: src/main/scala/com/reddit/recommender/ingestion/ScrapeConfig.java
//package com.reddit.recommender.ingestion;
//
//import java.util.ArrayList;
//import java.util.Arrays;
//import java.util.List;
//
//public class ScrapeConfig {
//    private List<String> subreddits;
//    private List<String> keywords;
//    private List<String> urls;
//    private int postsPerSource;
//    private int maxCommentsPerPost;
//    private boolean scrapeComments;
//    private String timeframe;
//    private String sort;
//    private boolean includeNsfw;
//    private int maxRetries;
//    private long delayBetweenSourcesMs;
//
//    private ScrapeConfig(Builder builder) {
//        this.subreddits = builder.subreddits;
//        this.keywords = builder.keywords;
//        this.urls = builder.urls;
//        this.postsPerSource = builder.postsPerSource;
//        this.maxCommentsPerPost = builder.maxCommentsPerPost;
//        this.scrapeComments = builder.scrapeComments;
//        this.timeframe = builder.timeframe;
//        this.sort = builder.sort;
//        this.includeNsfw = builder.includeNsfw;
//        this.maxRetries = builder.maxRetries;
//        this.delayBetweenSourcesMs = builder.delayBetweenSourcesMs;
//    }
//
//    // Builder class
//    public static class Builder {
//        private List<String> subreddits = new ArrayList<>();
//        private List<String> keywords = new ArrayList<>();
//        private List<String> urls = new ArrayList<>();
//        private int postsPerSource = 10;
//        private int maxCommentsPerPost = 5;
//        private boolean scrapeComments = false;
//        private String timeframe = "day";
//        private String sort = "relevance";
//        private boolean includeNsfw = false;
//        private int maxRetries = 3;
//        private long delayBetweenSourcesMs = 2000;
//
//        public Builder subreddits(String... subreddits) {
//            this.subreddits = Arrays.asList(subreddits);
//            return this;
//        }
//
//        public Builder subreddits(List<String> subreddits) {
//            this.subreddits = subreddits;
//            return this;
//        }
//
//        public Builder keywords(String... keywords) {
//            this.keywords = Arrays.asList(keywords);
//            return this;
//        }
//
//        public Builder urls(String... urls) {
//            this.urls = Arrays.asList(urls);
//            return this;
//        }
//
//        public Builder postsPerSource(int postsPerSource) {
//            if (postsPerSource < 10 || postsPerSource > 30) {
//                throw new IllegalArgumentException("postsPerSource must be between 10 and 30");
//            }
//            this.postsPerSource = postsPerSource;
//            return this;
//        }
//
//        public Builder maxCommentsPerPost(int maxCommentsPerPost) {
//            this.maxCommentsPerPost = maxCommentsPerPost;
//            return this;
//        }
//
//        public Builder scrapeComments(boolean scrapeComments) {
//            this.scrapeComments = scrapeComments;
//            return this;
//        }
//
//        public Builder timeframe(String timeframe) {
//            this.timeframe = timeframe;
//            return this;
//        }
//
//        public Builder sort(String sort) {
//            this.sort = sort;
//            return this;
//        }
//
//        public Builder includeNsfw(boolean includeNsfw) {
//            this.includeNsfw = includeNsfw;
//            return this;
//        }
//
//        public Builder maxRetries(int maxRetries) {
//            this.maxRetries = maxRetries;
//            return this;
//        }
//
//        public Builder delayBetweenSourcesMs(long delay) {
//            this.delayBetweenSourcesMs = delay;
//            return this;
//        }
//
//        public ScrapeConfig build() {
//            validate();
//            return new ScrapeConfig(this);
//        }
//
//        private void validate() {
//            if (subreddits.isEmpty() && keywords.isEmpty() && urls.isEmpty()) {
//                throw new IllegalStateException("At least one source (subreddits, keywords, or urls) must be specified");
//            }
//        }
//    }
//
//    // Getters
//    public List<String> getSubreddits() { return new ArrayList<>(subreddits); }
//    public List<String> getKeywords() { return new ArrayList<>(keywords); }
//    public List<String> getUrls() { return new ArrayList<>(urls); }
//    public int getPostsPerSource() { return postsPerSource; }
//    public int getMaxCommentsPerPost() { return maxCommentsPerPost; }
//    public boolean isScrapeComments() { return scrapeComments; }
//    public String getTimeframe() { return timeframe; }
//    public String getSort() { return sort; }
//    public boolean isIncludeNsfw() { return includeNsfw; }
//    public int getMaxRetries() { return maxRetries; }
//    public long getDelayBetweenSourcesMs() { return delayBetweenSourcesMs; }
//
//    // Factory methods for common configurations
//    // In ScrapeConfig.java, update trendingTechConfig():
//    public static ScrapeConfig trendingTechConfig() {
//        return new Builder()
//                .subreddits("programming", "technology", "gaming", "MachineLearning", "datascience")
//                .postsPerSource(10)     // Changed from 50 to 10
//                .timeframe("month")     // Changed from "day" to "month"
//                .sort("new")            // Changed from "hot" to "new"
//                .scrapeComments(true)   // Changed from false to true
//                .maxCommentsPerPost(1)  // Added this!
//                .build();
//    }
//
//    public static ScrapeConfig newsMonitoringConfig() {
//        return new Builder()
//                .keywords("news", "breaking", "update")
//                .subreddits("worldnews", "news", "politics")
//                .postsPerSource(100)
//                .timeframe("hour")
//                .sort("new")
//                .scrapeComments(true)
//                .maxCommentsPerPost(3)
//                .build();
//    }
//
//    public static ScrapeConfig competitorMonitoringConfig(String... competitorNames) {
//        return new Builder()
//                .keywords(competitorNames)
//                .postsPerSource(20)
//                .timeframe("week")
//                .sort("relevance")
//                .scrapeComments(true)
//                .maxCommentsPerPost(5)
//                .build();
//    }
//}

File: src/main/scala/com/reddit/recommender/ingestion/ScrapeConfig.scala
package com.reddit.recommender.ingestion

import scala.collection.JavaConverters._
import scala.collection.mutable.ListBuffer
import scala.annotation.varargs

class ScrapeConfig private (
                             private val subreddits: List[String],
                             private val keywords: List[String],
                             private val urls: List[String],
                             private val postsPerSource: Int,
                             private val maxCommentsPerPost: Int,
                             private val scrapeComments: Boolean,
                             private val timeframe: String,
                             private val sort: String,
                             private val includeNsfw: Boolean,
                             private val maxRetries: Int,
                             private val delayBetweenSourcesMs: Long
                           ) {
  def getSubreddits: java.util.List[String] = subreddits.asJava
  def getKeywords: java.util.List[String] = keywords.asJava
  def getUrls: java.util.List[String] = urls.asJava
  def getPostsPerSource: Int = postsPerSource
  def getMaxCommentsPerPost: Int = maxCommentsPerPost
  def isScrapeComments: Boolean = scrapeComments
  def getTimeframe: String = timeframe
  def getSort: String = sort
  def isIncludeNsfw: Boolean = includeNsfw
  def getMaxRetries: Int = maxRetries
  def getDelayBetweenSourcesMs: Long = delayBetweenSourcesMs
}

object ScrapeConfig {
  class Builder {
    private val subreddits = ListBuffer[String]()
    private val keywords = ListBuffer[String]()
    private val urls = ListBuffer[String]()
    private var postsPerSource: Int = 10
    private var maxCommentsPerPost: Int = 5
    private var scrapeComments: Boolean = false
    private var timeframe: String = "day"
    private var sort: String = "relevance"
    private var includeNsfw: Boolean = false
    private var maxRetries: Int = 3
    private var delayBetweenSourcesMs: Long = 2000L

    // Varargs version — matches Java: replaces the list
    @varargs
    def subreddits(subs: String*): Builder = {
      subreddits.clear()
      subreddits ++= subs
      this
    }

    @varargs
    def keywords(keys: String*): Builder = {
      keywords.clear()
      keywords ++= keys
      this
    }

    @varargs
    def urls(us: String*): Builder = {
      urls.clear()
      urls ++= us
      this
    }


    def subreddits(subs: java.util.List[String]): Builder = {
      subreddits.clear()
      subreddits ++= subs.asScala
      this
    }



    def postsPerSource(p: Int): Builder = {
      if (p < 10 || p > 30)
        throw new IllegalArgumentException("postsPerSource must be between 10 and 30")
      postsPerSource = p
      this
    }

    def maxCommentsPerPost(m: Int): Builder = { maxCommentsPerPost = m; this }
    def scrapeComments(s: Boolean): Builder = { scrapeComments = s; this }
    def timeframe(t: String): Builder = { timeframe = t; this }
    def sort(s: String): Builder = { sort = s; this }
    def includeNsfw(i: Boolean): Builder = { includeNsfw = i; this }
    def maxRetries(m: Int): Builder = { maxRetries = m; this }
    def delayBetweenSourcesMs(d: Long): Builder = { delayBetweenSourcesMs = d; this }

    def build(): ScrapeConfig = {
      if (subreddits.isEmpty && keywords.isEmpty && urls.isEmpty)
        throw new IllegalStateException("At least one source (subreddits, keywords, or urls) must be specified")

      new ScrapeConfig(
        subreddits.toList,
        keywords.toList,
        urls.toList,
        postsPerSource,
        maxCommentsPerPost,
        scrapeComments,
        timeframe,
        sort,
        includeNsfw,
        maxRetries,
        delayBetweenSourcesMs
      )
    }
  }

  def trendingTechConfig(): ScrapeConfig = new Builder()
    .subreddits("programming", "technology", "gaming", "MachineLearning", "datascience")
    .postsPerSource(10)
    .timeframe("month")
    .sort("new")
    .scrapeComments(true)
    .maxCommentsPerPost(1)
    .build()

  def newsMonitoringConfig(): ScrapeConfig = new Builder()
    .keywords("news", "breaking", "update")
    .subreddits("worldnews", "news", "politics")
    .postsPerSource(100)
    .timeframe("hour")
    .sort("new")
    .scrapeComments(true)
    .maxCommentsPerPost(3)
    .build()


  @varargs
  def competitorMonitoringConfig(competitorNames: String*): ScrapeConfig = new Builder()
    .keywords(competitorNames: _*)
    .postsPerSource(20)
    .timeframe("week")
    .sort("relevance")
    .scrapeComments(true)
    .maxCommentsPerPost(5)
    .build()
}